[JENKINS-28120] Bump the core requirement to Java7

kohsuke opened the pull request:
https://issues.jenkins-ci.org/browse/JENKINS-28120

As the first step of Java7 migration, ship Java7 class files without
requiring Java7 runtime.

Supercedes PR #1677 

@reviewbybees

kohsuke committed:
[JENKINS-28120]

As the first step of Java7 migration, ship Java7 class files without
requiring Java7 runtime.

jglick commented on changelog.html:
```diff
@@ -55,6 +55,10 @@
 <div id="trunk" style="display:none"><!--=TRUNK-BEGIN=-->
 <ul class=image>
+  <li class=rfe>
+    As <a hreef="http://jenkins-ci.org/content/good-bye-java6">promised</a>, shipping
```
Typo. Anyway we generally ask the merger, not the submitter, to add the `changelog.html` entry, given the high chance of conflicts.

jglick commented on pom.xml:
```diff
@@ -604,7 +604,7 @@ THE SOFTWARE.
           <signature>
             <groupId>org.codehaus.mojo.signature</groupId>
-            <artifactId>java1${java.level}</artifactId>
+            <artifactId>java16</artifactId>
```
Need a `TODO` comment explaining that it is intended to go back to `java1${java.level}` soon.

I would actually suggest leaving `java.level` alone and just temporarily overriding compiler configuration with `<target>1.7</target>`, so that we are still using `-source 6` until the decision is finalized.

kohsuke replied:
Enforcer would fail if we leave `${java.level}` as-is. If you really want to do it properly we'd have to create two variables. I don't think we need that, do we?

jglick replied:
I was not suggesting a new variable, just a temporary change like

```diff
diff --git a/pom.xml b/pom.xml
index 41cbe0d..57277a0 100644
--- a/pom.xml
+++ b/pom.xml
@@ -614,7 +614,7 @@ THE SOFTWARE.
         <artifactId>maven-compiler-plugin</artifactId>
         <configuration>
           <source>1.${java.level}</source>
-          <target>1.${java.level}</target>
+          <target>1.7</target> <!-- TODO restore to 1.${java.level} -->
           <!-- default reuseCreated is more performant 
           feel free to uncomment if you have any issues on your platform
           <compilerReuseStrategy>alwaysNew</compilerReuseStrategy>
@@ -639,7 +639,7 @@ THE SOFTWARE.
                   <version>3.0</version>
                 </requireMavenVersion>
                 <enforceBytecodeVersion>
-                  <maxJdkVersion>1.${java.level}</maxJdkVersion>
+                  <maxJdkVersion>1.7</maxJdkVersion> <!-- TODO restore to 1.${java.level} -->
                   <ignoreClasses>
                     <ignoreClass>org.eclipse.jetty.spdy.*</ignoreClass>
                   </ignoreClasses>
```

jglick replied:
The justification being that just as we do not want people using Java 7+ APIs in the near term, we do not want them writing new code using `-source 7` language constructs just yet.

jglick replied:
So if `java.level` is back to `6`, this diff line can and should be reverted.

jglick commented:
Prefer follow-on commits rather than rebasing, which messes up GitHub’s record of what was commented on when.

jglick commented:
IIRC there is a special main class somewhere which is compiled with `-source/target 5` that lets you know if you need to be using a newer JRE with a more polite message than a linkage error. Wherever that is, it should be updated to discuss Java 7 now.

kohsuke committed:
[JENKINS-28120]

Reflecting more code review comments from
https://github.com/jenkinsci/jenkins/pull/1679

- Jesse wants to keep the variable pointing to 1.6
- Update executable-war so that it fails gracefully if run with Java6 or
  earlier.
- CI build fails with VerifyError due to a regression in JDK.
  To support wide range of development environment, just cut off the
  verifier.
- Have enforcer require 1.7 although javac fails in a kinda obvious way.

stephenc commented:
:+1:

jglick commented:
For the benefit of reviewers, please include an actual summary in the PR title after the issue ID, and also include a hyperlink to the issue in the PR description.

jglick commented:
I also recommend pushing feature branches to a fork, not `origin`, unless you are expecting to collaborate with other developers on the branch:

* PR builders can test a new tip commit twice, if the job is configured to build against a wildcard branch. (Not the case for `jenkins` AFAICT.)
* The issue link daemon does not check that commits are in `master` before commenting in JIRA, resulting in confusing messages (as can be seen already in [JENKINS-28120](https://issues.jenkins-ci.org/browse/JENKINS-28120)) and even Fixed status.

kohsuke commented:
I actually disagree that the proposed change be pushed to a fork. Any proposed change should be open to collaboration with other developers in the community. For people blocked on issues, it's often useful to know that there's a PR that touches a specific issue.

My failure that resulted in the confusion in JENKINS-28120 is that I force-pushed this branch a few times. I should do that more carefully, for example perhaps only once when I eventually merge the PR.

jglick commented:
> Any proposed change should be open to collaboration with other developers in the community.

Well, they always are, if you are willing to file your own PR which just pulls commits from the original one. This is another thirty seconds of work and a potential source of confusion, so when you are _expecting_ multiple people to be pushing to the same branch/PR it is easier to do it in `origin`, but that is rarely the case in practice.

jglick commented:
BTW this was merged with my https://github.com/jenkinsci/jenkins/pull/1679#discussion_r29347245 unacknowledged.

KostyaSha commented:
And ssh-slaves should be also resolved https://issues.jenkins-ci.org/browse/JENKINS-28231
